cli_tools/gce_vm_image_import/importer/inflater.go

```diff
@@ -146,3 +294,146 @@ func getDisk(workflow *daisy.Workflow, diskIndex int) *daisy.Disk {
+	}
+	sourceFileSizeGb := (attrs.Size - 1) / 1073741824 + 1
+
+	diskUri := fmt.Sprintf("zones/%s/disks/%s", inflater.args.Zone, diskName)
+	pd := persistentDisk{
+		uri:           diskUri,
+		sizeGb:        cd.SizeGb,
+		sourceGb:      sourceFileSizeGb,
+		sourceType:    "vmdk", // only vmdk is supported right now
+		inflationTime: time.Since(startTime),
+	}
+
+	inflater.addTraceLog("Started checksum calculation.")
+	pd.checksum, err = inflater.calculateChecksum(ctx, diskUri)
+	return pd, err
+}
+
+// run a workflow to calculate checksum, and cleanup the disk
+func (inflater apiInflater) calculateChecksum(ctx context.Context, diskUri string) (string, error) {
+	w := daisy.New()
+	w.Name = "shadow-disk-checksum"
+	checksumScript := checksumScriptConst
+	w.Steps = map[string]*daisy.Step{
+		"create-disks": {
+			CreateDisks: &daisy.CreateDisks{
+				{
+					Disk: compute.Disk{
+						Name:        "disk-${NAME}",
+						SourceImage: "projects/compute-image-tools/global/images/family/debian-9-worker",
+						Type:        "pd-ssd",
+					},
+					FallbackToPdStandard: true,
+				},
+			},
+		},
+		"create-instance": {
+			CreateInstances: &daisy.CreateInstances{
+				Instances: []*daisy.Instance{
+					{
+						Instance: compute.Instance{
+							Name: "inst-${NAME}",
+							Disks: []*compute.AttachedDisk{
+								{Source: "disk-${NAME}"},
+								{Source: diskUri, Mode: "READ_ONLY"},
+							},
+							MachineType: "n1-highcpu-4",
+							Metadata: &compute.Metadata{
+								Items: []*compute.MetadataItems{
+									{
+										Key:   "startup-script",
+										Value: &checksumScript,
+									},
+								},
+							},
+							NetworkInterfaces: []*compute.NetworkInterface{
+								{
+									AccessConfigs: []*compute.AccessConfig{},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		"wait-for-checksum": {
+			WaitForInstancesSignal: &daisy.WaitForInstancesSignal{
+				{
+					Name: "inst-${NAME}",
+					SerialOutput: &daisy.SerialOutput{
+						Port:         1,
+						SuccessMatch: "Checksum calculated.",
+						StatusMatch: "Checksum:",
+					},
+				},
+			},
+		},
+	}
+	w.Dependencies = map[string][]string{
+		"create-instance": {"create-disks"},
+		"wait-for-checksum": {"create-instance"},
+	}
+
+	daisycommon.SetWorkflowAttributes(w, inflater.args.Project, inflater.args.Zone, inflater.args.ScratchBucketGcsPath,
+		inflater.args.Oauth, "20m", inflater.args.ComputeEndpoint, inflater.args.GcsLogsDisabled,
+		inflater.args.CloudLogsDisabled, inflater.args.StdoutLogsDisabled)
+	err := daisyUtils.RunWorkflowWithCancelSignal(ctx, w)
+	if err != nil {
+		return "", err
+	}
+	return w.GetSerialConsoleOutputValue("disk-checksum"), nil
+}
+
+const (
+	checksumScriptConst = `
+		function serialOutputKeyValuePair() {
+			echo "<serial-output key:'$1' value:'$2'>"
+		}
+
+		CHECK_DEVICE=sdb
+		BLOCK_COUNT=$(cat /sys/class/block/$CHECK_DEVICE/size)
+	
+		# Check size = 200000*512 = 100MB
+		CHECK_COUNT=200000
```

Comment:
This means that you calculate the checksum of the first 100MB?

Reply:
As introduced in the design doc, we calculate 0~0.1GB, 9.9GB~10GB, 99.9GB~100GB, and max-0.1GB ~ max GB

Reply:
Since shadow logic will not be used in customer projects, is it feasible to run calculate the full checksum on whole disks, since we're paying for it and any time penalty is not a big issue (unless it takes many hours to calculate)?

Reply:
It will be used by customer in stage 2. And the penalty is high - as high as around 50% of inflating the disk in traditional way.

Reply:
Can we do the full disk checksum calculation for our internal tests and this faster logic for external customers? We should not worry about the extra time or cost in our internal projects and this will give us much higher confidence VMDK API works correctly.

Reply:
I have manually tested full checksum for those internal tests used images.

Reply:
It would be still useful to run full checksum calculation, in case VMDK API is flaky (returns correct PD one time, not the next time) which can't be always caught by one-time manual test.

Reply:
Talked offline and agreed to run full checksum tests locally in a loop for a couple of days